File: app/src/main/java/ss/nscube/webshare/server/utils/FileUtil.java
package ss.nscube.webshare.server.utils;

import android.annotation.SuppressLint;
import android.util.Log;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

public class FileUtil {
    private static Map<String, String> mimeTypes;

    static {

        mimeTypes = new HashMap<String, String>(); 
        mimeTypes.put("epub", "application/epub+zip");
        mimeTypes.put("gz", "application/gzip");
        mimeTypes.put("jar", "application/java-archive");
        mimeTypes.put("ser", "application/java-serialized-object");
        mimeTypes.put("class", "application/java-vm");
        mimeTypes.put("js", "application/javascript");
        mimeTypes.put("json", "application/json");
        mimeTypes.put("doc", "application/msword");
        mimeTypes.put("dot", "application/msword");
        mimeTypes.put("bin", "application/octet-stream");
        mimeTypes.put("deploy", "application/octet-stream");
        mimeTypes.put("msu", "application/octet-stream");
        mimeTypes.put("msp", "application/octet-stream");
        mimeTypes.put("pdf", "application/pdf");
        mimeTypes.put("prf", "application/pics-rules");
        mimeTypes.put("xhtml", "application/xhtml+xml");
        mimeTypes.put("xht", "application/xhtml+xml");
        mimeTypes.put("xml", "application/xml");
        mimeTypes.put("xsd", "application/xml");
        mimeTypes.put("xsl", "application/xslt+xml");
        mimeTypes.put("xslt", "application/xslt+xml");
        mimeTypes.put("xspf", "application/xspf+xml");
        mimeTypes.put("zip", "application/zip");
        mimeTypes.put("apk", "application/vnd.android.package-archive");
        mimeTypes.put("deb", "application/vnd.debian.binary-package");
        mimeTypes.put("ddeb", "application/vnd.debian.binary-package");
        mimeTypes.put("udeb", "application/vnd.debian.binary-package");
        mimeTypes.put("torrent", "application/x-bittorrent");
        mimeTypes.put("7z", "application/x-7z-compressed");
        mimeTypes.put("info", "application/x-info");
        mimeTypes.put("jnlp", "application/x-java-jnlp-file");
        mimeTypes.put("m3u8", "application/x-mpegURL");
        mimeTypes.put("tar", "application/x-tar");
        mimeTypes.put("flac", "audio/flac");
        mimeTypes.put("mid", "audio/midi");
        mimeTypes.put("midi", "audio/midi");
        mimeTypes.put("kar", "audio/midi");
        mimeTypes.put("mpga", "audio/mpeg");
        mimeTypes.put("mpega", "audio/mpeg");
        mimeTypes.put("mp2", "audio/mpeg");
        mimeTypes.put("mp3", "audio/mpeg");
        mimeTypes.put("m4a", "audio/mpeg");
        mimeTypes.put("m3u", "audio/mpegurl");
        mimeTypes.put("oga", "audio/ogg");
        mimeTypes.put("ogg", "audio/ogg");
        mimeTypes.put("opus", "audio/ogg");
        mimeTypes.put("spx", "audio/ogg");
        mimeTypes.put("wav", "audio/x-wav");
        mimeTypes.put("ttc", "font/collection");
        mimeTypes.put("otf", "font/otf");
        mimeTypes.put("ttf", "font/ttf");
        mimeTypes.put("woff", "font/woff");
        mimeTypes.put("woff2", "font/woff2");
        mimeTypes.put("gif", "image/gif");
        mimeTypes.put("ief", "image/ief");
        mimeTypes.put("jp2", "image/jp2");
        mimeTypes.put("jpg2", "image/jp2");
        mimeTypes.put("jpeg", "image/jpeg");
        mimeTypes.put("jpg", "image/jpeg");
        mimeTypes.put("jpe", "image/jpeg");
        mimeTypes.put("jpm", "image/jpm");
        mimeTypes.put("jpx", "image/jpx");
        mimeTypes.put("jpf", "image/jpx");
        mimeTypes.put("pcx", "image/pcx");
        mimeTypes.put("png", "image/png");
        mimeTypes.put("svg", "image/svg+xml");
        mimeTypes.put("svgz", "image/svg+xml");
        mimeTypes.put("tiff", "image/tiff");
        mimeTypes.put("tif", "image/tiff");
        mimeTypes.put("css", "text/css");
        mimeTypes.put("html", "text/html");
        mimeTypes.put("htm", "text/html");
        mimeTypes.put("shtml", "text/html");
        mimeTypes.put("asc", "text/plain");
        mimeTypes.put("txt", "text/plain");
        mimeTypes.put("text", "text/plain");
        mimeTypes.put("pot", "text/plain");
        mimeTypes.put("brf", "text/plain");
        mimeTypes.put("srt", "text/plain");
        mimeTypes.put("h++", "text/x-c++hdr");
        mimeTypes.put("hpp", "text/x-c++hdr");
        mimeTypes.put("hxx", "text/x-c++hdr");
        mimeTypes.put("hh", "text/x-c++hdr");
        mimeTypes.put("c++", "text/x-c++src");
        mimeTypes.put("cpp", "text/x-c++src");
        mimeTypes.put("cxx", "text/x-c++src");
        mimeTypes.put("cc", "text/x-c++src");
        mimeTypes.put("h", "text/x-chdr");
        mimeTypes.put("htc", "text/x-component");
        mimeTypes.put("csh", "text/x-csh");
        mimeTypes.put("c", "text/x-csrc");
        mimeTypes.put("d", "text/x-dsrc");
        mimeTypes.put("diff", "text/x-diff");
        mimeTypes.put("patch", "text/x-diff");
        mimeTypes.put("hs", "text/x-haskell");
        mimeTypes.put("java", "text/x-java");
        mimeTypes.put("ly", "text/x-lilypond");
        mimeTypes.put("lhs", "text/x-literate-haskell");
        mimeTypes.put("py", "text/x-python");
        mimeTypes.put("scala", "text/x-scala");
        mimeTypes.put("mp4", "video/mp4");
        mimeTypes.put("qt", "video/quicktime");
        mimeTypes.put("mov", "video/quicktime");
        mimeTypes.put("mpeg", "video/mpeg");
        mimeTypes.put("mpg", "video/mpeg");
        mimeTypes.put("mpe", "video/mpeg");
        mimeTypes.put("webm", "video/webm");
        mimeTypes.put("mpv", "video/x-matroska");
        mimeTypes.put("mkv", "video/x-matroska");
        mimeTypes.put("avi", "video/x-msvideo");

    }

    public static String getMIMEType(String name) {
        String extension = name.substring(name.lastIndexOf(".") + 1).toLowerCase();
        if (mimeTypes.containsKey(extension)) {
            return mimeTypes.get(extension);
        } else {
            return "application/octet-stream";
        }
    }

    public static String getMIMETypeOnly(String name) {
        String extension = name.substring(name.lastIndexOf(".") + 1).toLowerCase();
        if (mimeTypes.containsKey(extension)) {
            return mimeTypes.get(extension);
        } else {
            return null;
        }
    }

    public static String getMIMEType(File file) {
        String extension = file.getName().substring(file.getName().lastIndexOf(".") + 1).toLowerCase();
        if (mimeTypes.containsKey(extension)) {
            return mimeTypes.get(extension);
        } else {
            return "application/octet-stream";
        }
    }

    public static String getType(File file) {
        String type = "file";
        if (file.isFile()) {
            if (file.getName().endsWith(".text")) {
                return "text";
            }
            String mimeType = getMIMEType(file);
            if (mimeType == null) {
                 type = "file";
            } else if (mimeType.contains("audio"))
                type = "audio";
            else if (mimeType.contains("image"))
                type = "image";
            else if (mimeType.contains("video"))
                type = "video";
            return type;
        } else if (file.isDirectory()) {
            type = "folder";
        }
        return type;
    }

    @SuppressLint("DefaultLocale")
    public static String getSize(long size) {
        long kilo = 1024;
        long mega = kilo * kilo;
        long giga = mega * kilo;
        long tera = giga * kilo;
        String s = "";
        double kb = (double)size / kilo;
        double mb = kb / kilo;
        double gb = mb / kilo;
        double tb = gb / kilo;
        if(size < kilo) {
            s = size + " Bytes";
        } else if(size >= kilo && size < mega) {
            s =  String.format("%.2f", kb) + " KB";
        } else if(size >= mega && size < giga) {
            s = String.format("%.2f", mb) + " MB";
        } else if(size >= giga && size < tera) {
            s = String.format("%.2f", gb) + " GB";
        } else if(size >= tera) {
            s = String.format("%.2f", tb) + " TB";
        }
        return s;
    }

    public static boolean isAcceptable(String fileName) {
        int slashPos = fileName.indexOf('/');
        Log.i("TAG", "isAcceptable: " + fileName + " " + slashPos);
        return slashPos == -1;
    }
}

File: app/src/main/java/ss/nscube/webshare/server/utils/FileUtil.kt
package ss.nscube.webshare.server.utils

import android.annotation.SuppressLint
import ss.nscube.webshare.utils.WebFileUtil
import ss.nscube.webshare.utils.log
import java.io.IOException
import java.io.OutputStream
import java.util.*

object FileUtil {
    const val OctetStream =  "application/octet-stream"
    const val JsonMimeType =  "application/json"

    private var mimeTypes: MutableMap<String, String> = HashMap()

    fun getExtension(fileName: String): String? {
        val lastDotIndex = fileName.lastIndexOf('.')
        if (lastDotIndex == -1) return null
        val nameWithoutExtension = fileName.substring(0, lastDotIndex)
        val lastSecondDotIndex = nameWithoutExtension.lastIndexOf('.')
        val extension = fileName.substring(lastDotIndex + 1).lowercase(Locale.getDefault())
        if (extension.isEmpty()) return null
        if (lastSecondDotIndex < 1) return extension
        val secondExtension = nameWithoutExtension.substring(lastSecondDotIndex + 1).lowercase(Locale.getDefault())
        if (secondExtension == "tar") return "$secondExtension.$extension"
        return extension
    }

    fun getMimeTypeFromName(name: String): String {
        val extension = name.substring(name.lastIndexOf(".") + 1).lowercase(Locale.getDefault())
        val mime = getMimeTypeFromExtension(extension)
        log("EXTENSION $mime $extension")
        return mime
    }

    fun getMimeTypeFromExtension(extension: String?): String {
        return mimeTypes[extension] ?: OctetStream
    }

    fun getFileType(name: String): String {
        val extension = name.substring(name.lastIndexOf(".") + 1).lowercase(Locale.getDefault())
        val mime = mimeTypes[extension] ?: OctetStream
        return getFileType(extension, mime)
    }

    fun getFileType(extension: String?, mime: String): String {
        return when {
            mime.startsWith(WebFileUtil.Image) -> WebFileUtil.Image
            mime.startsWith(WebFileUtil.Video) -> WebFileUtil.Video
            mime.startsWith(WebFileUtil.Audio) -> WebFileUtil.Audio
            extension.equals("apk", true) -> WebFileUtil.App
            else -> WebFileUtil.Document
        }
    }

    val kilo: Long = 1024
    val mega = kilo * kilo
    val giga = mega * kilo
    val tera = giga * kilo

    @SuppressLint("DefaultLocale")
    fun getSize(size: Long): String {
        var s = ""
        val kb = size.toDouble() / kilo
        val mb = kb / kilo
        val gb = mb / kilo
        val tb = gb / kilo
        s = if (size < kilo) {
            "$size Bytes"
        } else if (size in kilo until mega) {
            formatSize(size, kilo, kb) + " KB"
        } else if (size in mega until giga) {
            formatSize(size, mega, mb) + " MB"
        } else if (size in giga until tera) {
            formatSize(size, giga, gb) + " GB"
        } else {
            formatSize(size, tera, tb) + " TB"
        }
        return s
    }

    fun formatSize(length: Long, decimalDivisible: Long, size: Double): String {
        val isWholeNumber = length % decimalDivisible == 0L
        return String.format(if (isWholeNumber) "%d" else "%.2f", if (isWholeNumber) size.toInt() else size)
    }

    fun nullOutputStream(): OutputStream {
        return object : OutputStream() {
            @Volatile
            private var closed = false
            @Throws(IOException::class)
            private fun ensureOpen() {
                if (closed) {
                    throw IOException("Stream closed")
                }
            }

            @Throws(IOException::class)
            override fun write(b: Int) {
                ensureOpen()
            }

            @Throws(IOException::class)
            override fun write(b: ByteArray, off: Int, len: Int) {
                Objects.checkFromIndexSize(off, len, b.size)
                ensureOpen()
            }

            override fun close() {
                closed = true
            }
        }
    }

    init {
        mimeTypes["epub"] = "application/epub+zip"
        mimeTypes["gz"] = "application/gzip"
        mimeTypes["jar"] = "application/java-archive"
        mimeTypes["ser"] = "application/java-serialized-object"
        mimeTypes["class"] = "application/java-vm"
        mimeTypes["js"] = "application/javascript"
        mimeTypes["json"] = "application/json"
        mimeTypes["doc"] = "application/msword"
        mimeTypes["dot"] = "application/msword"
        mimeTypes["bin"] = "application/octet-stream"
        mimeTypes["deploy"] = "application/octet-stream"
        mimeTypes["msu"] = "application/octet-stream"
        mimeTypes["msp"] = "application/octet-stream"
        mimeTypes["pdf"] = "application/pdf"
        mimeTypes["prf"] = "application/pics-rules"
        mimeTypes["xhtml"] = "application/xhtml+xml"
        mimeTypes["xht"] = "application/xhtml+xml"
        mimeTypes["xml"] = "application/xml"
        mimeTypes["xsd"] = "application/xml"
        mimeTypes["xsl"] = "application/xslt+xml"
        mimeTypes["xslt"] = "application/xslt+xml"
        mimeTypes["xspf"] = "application/xspf+xml"
        mimeTypes["zip"] = "application/zip"
        mimeTypes["apk"] = "application/octet-stream"
        mimeTypes["deb"] = "application/vnd.debian.binary-package"
        mimeTypes["ddeb"] = "application/vnd.debian.binary-package"
        mimeTypes["udeb"] = "application/vnd.debian.binary-package"
        mimeTypes["torrent"] = "application/x-bittorrent"
        mimeTypes["7z"] = "application/x-7z-compressed"
        mimeTypes["info"] = "application/x-info"
        mimeTypes["jnlp"] = "application/x-java-jnlp-file"
        mimeTypes["m3u8"] = "application/x-mpegURL"
        mimeTypes["tar"] = "application/x-tar"
        mimeTypes["aac"] = "audio/aac"
        mimeTypes["flac"] = "audio/flac"
        mimeTypes["mid"] = "audio/midi"
        mimeTypes["midi"] = "audio/midi"
        mimeTypes["kar"] = "audio/midi"
        mimeTypes["mpga"] = "audio/mpeg"
        mimeTypes["mpega"] = "audio/mpeg"
        mimeTypes["mp2"] = "audio/mpeg"
        mimeTypes["mp3"] = "audio/mpeg"
        mimeTypes["m4a"] = "audio/mpeg"
        mimeTypes["m3u"] = "audio/mpegurl"
        mimeTypes["oga"] = "audio/ogg"
        mimeTypes["ogg"] = "audio/ogg"
        mimeTypes["opus"] = "audio/ogg"
        mimeTypes["spx"] = "audio/ogg"
        mimeTypes["wav"] = "audio/x-wav"
        mimeTypes["ttc"] = "font/collection"
        mimeTypes["otf"] = "font/otf"
        mimeTypes["ttf"] = "font/ttf"
        mimeTypes["woff"] = "font/woff"
        mimeTypes["woff2"] = "font/woff2"
        mimeTypes["gif"] = "image/gif"
        mimeTypes["ief"] = "image/ief"
        mimeTypes["jp2"] = "image/jp2"
        mimeTypes["jpg2"] = "image/jp2"
        mimeTypes["jpeg"] = "image/jpeg"
        mimeTypes["jpg"] = "image/jpeg"
        mimeTypes["jpe"] = "image/jpeg"
        mimeTypes["jpm"] = "image/jpm"
        mimeTypes["jpx"] = "image/jpx"
        mimeTypes["jpf"] = "image/jpx"
        mimeTypes["pcx"] = "image/pcx"
        mimeTypes["png"] = "image/png"
        mimeTypes["svg"] = "image/svg+xml"
        mimeTypes["svgz"] = "image/svg+xml"
        mimeTypes["tiff"] = "image/tiff"
        mimeTypes["tif"] = "image/tiff"
        mimeTypes["css"] = "text/css"
        mimeTypes["html"] = "text/html"
        mimeTypes["htm"] = "text/html"
        mimeTypes["shtml"] = "text/html"
        mimeTypes["asc"] = "text/plain"
        mimeTypes["txt"] = "text/plain"
        mimeTypes["text"] = "text/plain"
        mimeTypes["pot"] = "text/plain"
        mimeTypes["brf"] = "text/plain"
        mimeTypes["srt"] = "text/plain"
        mimeTypes["h++"] = "text/x-c++hdr"
        mimeTypes["hpp"] = "text/x-c++hdr"
        mimeTypes["hxx"] = "text/x-c++hdr"
        mimeTypes["hh"] = "text/x-c++hdr"
        mimeTypes["c++"] = "text/x-c++src"
        mimeTypes["cpp"] = "text/x-c++src"
        mimeTypes["cxx"] = "text/x-c++src"
        mimeTypes["cc"] = "text/x-c++src"
        mimeTypes["h"] = "text/x-chdr"
        mimeTypes["htc"] = "text/x-component"
        mimeTypes["csh"] = "text/x-csh"
        mimeTypes["c"] = "text/x-csrc"
        mimeTypes["d"] = "text/x-dsrc"
        mimeTypes["diff"] = "text/x-diff"
        mimeTypes["patch"] = "text/x-diff"
        mimeTypes["hs"] = "text/x-haskell"
        mimeTypes["java"] = "text/x-java"
        mimeTypes["ly"] = "text/x-lilypond"
        mimeTypes["lhs"] = "text/x-literate-haskell"
        mimeTypes["py"] = "text/x-python"
        mimeTypes["sh"] = "text/x-shellscript"
        mimeTypes["scala"] = "text/x-scala"
        mimeTypes["mp4"] = "video/mp4"
        mimeTypes["qt"] = "video/quicktime"
        mimeTypes["mov"] = "video/quicktime"
        mimeTypes["mpeg"] = "video/mpeg"
        mimeTypes["mpg"] = "video/mpeg"
        mimeTypes["mpe"] = "video/mpeg"
        mimeTypes["webm"] = "video/webm"
        mimeTypes["mpv"] = "video/x-matroska"
        mimeTypes["mkv"] = "video/x-matroska"
        mimeTypes["avi"] = "video/x-msvideo"
    }
}
